lib/mongoid/association/referenced/has_one/proxy.rb

```diff
@@ -49,17 +49,14 @@ def nullify
           #
           # @return [ One ] The association.
           def substitute(replacement)
-            # If the same object currently associated is being assigned,
-            # rebind the association and save the target but do not destroy
-            # the target.
-
-            unbind_one
-            if persistable?
-              # TODO can this entire method be skipped if self == replacement?
-              if _association.destructive? && self != replacement
-                send(_association.dependent)
-              else
-                save if persisted?
+            if replacement != self
```

Comment:
There is precedent for this in the EmbedsOne proxy

Reply:
I think this is fine except the comparison should be for identity (via `equal?`), not equality (via `==`). Looking at `==` implementation in `lib/mongoid/equality.rb`, two objects appear to compare equal if they have the same ids, but in a case like that (which I am hoping my other comment  would reproduce) the rebinding is in fact necessary on both objects.

Assuming my hypothesis is correct, the other place which performs `self != replacement` comparison similarly would need to be changed to use identity comparison.